plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/perspective/terminal/TerminalPresenter.java

```diff
@@ -164,6 +168,22 @@ private void connectToTerminalWebSocket(@NotNull String wsUrl) {
 
         socket = WebSocket.create(wsUrl);
 
+        socket.setOnMessageHandler(new MessageReceivedHandler() {
+            @Override
+            public void onMessageReceived(MessageReceivedEvent event) {
+                terminal.write(event.getMessage());
+            }
+        });
+
+        socket.setOnCloseHandler(new ConnectionClosedHandler() {
+            @Override
+            public void onClose(WebSocketClosedEvent event) {
+                if (CLOSE_NORMAL == event.getCode()) {
```

Comment:
Because, in case abnormal closing launches onErrorHandler and we try reconnect to terminal...